pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+	if ttlExpired, remainingTime := reconcileRequired(&dnsNameResolver.Status); ttlExpired {
+		return reconcile.Result{Requeue: true, RequeueAfter: remainingTime}, nil
+	}
+
+	dnsName := string(dnsNameResolver.Spec.Name)
+	matchesRegular := !isWildcard(dnsName)
+
+	// Add the DNS name in the spec of the DNSNameResolver resource to the resolver with empty resolved addresses. This is done
+	// to ensure that the resolver sends the resolution request for the DNS name if it is getting added for the first time.
+	r.res.Add(dnsName, []networkv1alpha1.DNSNameResolverResolvedAddress{}, matchesRegular, request.Name)
+
+	// Iterate through each of the resolved names matching the DNS name and add them to the resolver along with the current IP addresses.
+	for _, resolvedName := range dnsNameResolver.Status.ResolvedNames {
+		r.res.Add(string(resolvedName.DNSName), resolvedName.ResolvedAddresses, matchesRegular, request.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// isWildcard checks if the domain name is wildcard.
+func isWildcard(dnsName string) bool {
+	trimmedString := strings.TrimPrefix(dnsName, "*")
+	return trimmedString != dnsName
+}
+
+// removalOfIPsRequired checks whether the TTL of any IP address, associated to a DNS name
+// matching the DNSNameResolver object, has expired and the grace period after the TTL
+// expiration is also over. If so, then those resolved addresses are removed from the
+// resolved name details of the DNS name.
+func removalOfIPsRequired(status *networkv1alpha1.DNSNameResolverStatus) bool {
+	updated := false
+	for index, resolvedName := range status.ResolvedNames {
+		updateResolvedAddresses := []networkv1alpha1.DNSNameResolverResolvedAddress{}
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			if time.Now().Before(resolvedAddress.LastLookupTime.Time.Add(time.Second*time.Duration(resolvedAddress.TTLSeconds) + ipRemovalGracePeriod)) {
+				updateResolvedAddresses = append(updateResolvedAddresses, resolvedAddress)
+			} else {
+				updated = true
+			}
+		}
+		status.ResolvedNames[index].ResolvedAddresses = updateResolvedAddresses
+	}
+	return updated
+}
+
+// reconcileRequired checks whether the TTL of any IP address, associated to a DNS name
+// matching the DNSNameResolver object, has expired. If so, then the minimum remaining
+// time until the grace period gets over among that of the IP addresses with expired
+// TTLs is returned.
+func reconcileRequired(status *networkv1alpha1.DNSNameResolverStatus) (bool, time.Duration) {
+	var minRemainingTime time.Duration
+	ttlExpired := false
+	for _, resolvedName := range status.ResolvedNames {
+		for _, addr := range resolvedName.ResolvedAddresses {
+			if !time.Now().Before(addr.LastLookupTime.Time.Add(time.Second * time.Duration(addr.TTLSeconds))) {
+				remainingTime := time.Until(addr.LastLookupTime.Time.Add(time.Second*time.Duration(addr.TTLSeconds) + ipRemovalGracePeriod))
```

Comment:
Is it possible for `remainingTime` to be negative here, or is it guaranteed that any resolved addresses past the grace period are cleaned up elsewhere?

Reply:
To answer my own question, looks like `removalOfIPsRequired` is called before `reconcileRequired`, and it removes any already expired addresses, so this shouldn't be an issue

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2